### DISCUSSION TITLE
Best way to pass dependencies to other functions?

### DISCUSSION THREAD

**Discussion by HenryDashwood** [fastapi/fastapi#6288] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
.
```

### Description

If I have a situation like below I will get an error because you can't use Depends() with non-route functions it seems. Specifically, I will get an error saying `AttributeError: 'Depends' object has no attribute 'exec'`

```python
def get_session():
    with Session(engine) as session:
        yield session

def some_other_func(*, session: Session = Depends(get_session)):
    users = session.exec(select(User))
    print("doing some other func thing")

@router.post("/demo")
def demo(*, session: Session = Depends(get_session)):
    some_other_func()
    return {"demo": "success"}
```

If instead, I try to pass the session object from one function to another as an argument, FastAPI will also complain. Specifically, it says `TypeError: 'ModelMetaclass' object is not iterable`

```python
def get_session():
    with Session(engine) as session:
        yield session

def some_other_func(session: Session):
    users = session.exec(select(User))
    print("doing some other func thing")

@router.post("/demo")
def demo(*, session: Session = Depends(get_session)):
    some_other_func(session)
    return {"demo": "success"}
```

What's the best way of giving a non-route function access to the database session? If someone could also explain why non-route functions _are_ able to make use of `Depends` in [the security tutorial](https://fastapi.tiangolo.com/tutorial/security/oauth2-jwt/) I'd be very grateful

### Operating System

macOS

### FastAPI Version

0.79.0

### Python Version

3.10.3

**Comment by JarroVGIT** [comment, marked as answer]
FastAPI performs its magic only when a function paramater with a default value as an instance of `Depends` is called in the context of a request. You can chain them, and FastAPI will resolve them recursivly. The resolvement of all `Depends` instances does **not** happen in the `Depends` class itself, but happens from the outside when FastAPI determines what parameters are expected or which functions must be executed when a request hits a certain endpoint.

So to answer your first question, the best way to provide a 'normal' function with a variable that is originally injected through a `Depends()` parameter is indeed like you have:
```python
@router.post("/demo")
def demo(*, session: Session = Depends(get_session)):
    some_other_func(session)
    return {"demo": "success"}
```

To answer your second question (why does this work in the security tutorial): all functions that use `Depends()` are depending on each other. You can 'chain' functions that use `Depends()` as their parameters and FastAPI will resolve them all **in the context of a request**.

**Comment by HenryDashwood** [comment]
Thank you. This did help clarify my thinking and I've solved the issue
